pkg/scheduler/queue/tree_queue.go

```diff
@@ -0,0 +1,282 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package queue
+
+import (
+	"container/list"
+)
+
+type QueuePath []string //nolint:revive
+type QueueIndex int     //nolint:revive
+
+const localQueueIndex = -1
+
+// TreeQueue is a hierarchical queue implementation with an arbitrary amount of child queues.
+//
+// TreeQueue internally maintains round-robin fair queuing across all of its queue dimensions.
+// Each queuing dimension is modeled as a node in the tree, internally reachable through a QueuePath.
+//
+// The QueuePath is an ordered array of strings describing the path through the tree to the node,
+// which contains the FIFO local queue of all items enqueued for that queuing dimension.
+//
+// When dequeuing from a given node, the node will round-robin equally between dequeuing directly
+// from its own local queue and dequeuing recursively from its list of child TreeQueues.
+// No queue at a given level of the tree is dequeued from consecutively unless all others
+// at the same level of the tree are empty down to the leaf node.
+type TreeQueue struct {
+	// name of the tree node will be set to its segment of the queue path
+	name                   string
+	maxQueueLen            int
+	localQueue             *list.List
+	currentChildQueueIndex int
+	childQueueOrder        []string
+	childQueueMap          map[string]*TreeQueue
+}
+
+func NewTreeQueue(name string, maxQueueLen int) *TreeQueue {
+	return &TreeQueue{
+		name:                   name,
+		maxQueueLen:            maxQueueLen,
+		localQueue:             list.New(),
+		currentChildQueueIndex: localQueueIndex,
+		childQueueMap:          map[string]*TreeQueue{},
+		childQueueOrder:        nil,
+	}
+}
+
+func (q *TreeQueue) IsEmpty() bool {
+	// avoid recursion to make this a cheap operation
+	//
+	// Because we dereference empty child nodes during dequeuing,
+	// we assume that emptiness means there are no child nodes
+	// and nothing in this tree node's local queue.
+	//
+	// In reality a package member could attach empty child queues with getOrAddNode
+	// in order to get a functionally-empty tree that would report false for IsEmpty.
+	// We assume this does not occur or is not relevant during normal operation.
+	return q.localQueue.Len() == 0 && len(q.childQueueMap) == 0
+}
+
+func (q *TreeQueue) NodeCount() int {
```

Comment:
[nit] Perhaps `SubtreeSize()` (or something like that) would be a clearer name? When I saw `NodeCount()` I assumed this was referring only to immediate children.

Reply:
I would prefer to leave it as is, `NodeCount` is a standard name for this operation on trees across data structures and algorithm textbooks & other reference materials. "Size" is not a consistently defined term across tree types, as sometimes it does not include the root node.

The Tree Queue requires even clearer disambiguation, as "size" could also refer to the queue size. Using "subtree" also doesn't make it clear what is being referred to, as we don't know whether the node itself is included in the term subtree, and the node (not including itself) has 0 to N subtrees, not 1 subtree.

Reply:
>I will add docstring to clarify.
Perfect, thanks.